### DISCUSSION TITLE
Scrolling text: how to do it using an RTOS?

### DISCUSSION THREAD

**Discussion by dolence** [olikraus/u8g2#2130] [Q&A] [answered]
Hi! I'm scratching my head trying to figure this out...

This is the scrolling text example for Arduino:

```
void loop(void) {
  u8g2_uint_t x;
  
  u8g2.firstPage();
  do {
  
    // draw the scrolling text at current offset
    x = offset;
    u8g2.setFont(u8g2_font_inb30_mr);		// set the target font
    do {								// repeated drawing of the scrolling text...
      u8g2.drawUTF8(x, 30, text);			// draw the scolling text
      x += width;						// add the pixel width of the scrolling text
    } while( x < u8g2.getDisplayWidth() );		// draw again until the complete display is filled
    
    u8g2.setFont(u8g2_font_inb16_mr);		// draw the current pixel width
    u8g2.setCursor(0, 58);
    u8g2.print(width);					// this value must be lesser than 128 unless U8G2_16BIT is set
    
  } while ( u8g2.nextPage() );
  
  offset-=1;							// scroll by one pixel
  if ( (u8g2_uint_t)offset < (u8g2_uint_t)-width )	
    offset = 0;							// start over again
    
  delay(10);							// do some small delay
}
```

I want to scroll a text string in a more RTOS way. I'm using STM32 with HAL libraries, FreeRTOS and the C version of U8G2.
I have a freertos.c file where I start a dedicated task which draw a simple status bar and updates the display.

```
void StartDisp(void *argument)
{
  /* USER CODE BEGIN StartDisp */

  /* Infinite loop */
  for(;;)
  {

	  u8g2_FirstPage(&u8g2);
	  do {
		  draw_bar();
		  draw_menu();
	  } while (u8g2_NextPage(&u8g2));

	  osDelay(10);
  }
  /* USER CODE END StartDisp */
}
```

I also have gui.c and gui.h where the drawing functions are being written to keep the code organized. What I haven't figured out is how to implement that Arduino loop in a non blocking way. How to update offset and new x position without locking everything on do/while loops and considering page buffer switching occurs in another file?

```
void scroll_text(u8g2_t *u8g2, u8g2_uint_t x, u8g2_uint_t y, const char *str, const uint8_t  *font)
{
	//TODO  
}

void draw_menu()
{
	u8g2_DrawRBox(&u8g2, 46, 27, 79, 25, 3);

	u8g2_SetDrawColor(&u8g2, 0);
	u8g2_SetFont(&u8g2, u8g2_font_nokiafc22_tu);
	u8g2_DrawUTF8(&u8g2, 48, 38, "Test");

	scroll_text(&u8g2, 48, 45, "This is just a simple test", u8g2_font_5x7_tf); 

	u8g2_SetDrawColor(&u8g2, 1);
}
```

Any help would be very appreciated.

**Comment by olikraus** [comment]
I do not know much about your problem, but I suggest two improvements:

1. Introduce a "isUpdate" flag. 

Only if the isUpdate flag is true (1), then redraw the display content. It might look like this:

``` C++
volatile int isUpdate = 1;  // init with true so that display is updated during startup

void StartDisp(void *argument)
{
  /* USER CODE BEGIN StartDisp */

  /* Infinite loop */
  for(;;)
  {
          while( isUpdate == 0 )  // do not update the display as long as isUpdate is false
            ;
          /* isUpdate is not 0, so do the update */
	  u8g2_FirstPage(&u8g2);
	  do {
		  draw_bar();
		  draw_menu();
	  } while (u8g2_NextPage(&u8g2));
          isUpdate = 0; // display update is done, clear the isUpdate flag and wait until it gets set again
  }
  /* USER CODE END StartDisp */
}
```

Obviously you need the "volatile" attribute to avoid any compiler caching of the variable.

Now you can set the isUpdate flag, whenever something has changed for the display, for example an offset for the text (scroll effect) or any update of the menu after any button press event.

The main benefit is to reduce load on the uC. It also helps to react more quickly on any changes.

2. Create shadow variables for drawing

U8g2 picture loop requires constant values. Especially if you have multiple parallel tasks, modifing a variable while updating the display will lead to visible artefacts.
Let me assume, that "so" is the scroll offset, which is used in you main task:

``` C++
void main_task() {
  for() {
    so+=2; // move text
    isUpdate = 1;
    task_delay( 100 milliseconds );
  }
}
```
Then your draw procedure should work on a shadow variable (lets call it "soShadow"):


``` C++
void draw_menu() {
	u8g2_DrawRBox(&u8g2, 46, 27, 79, 25, 3);

	u8g2_SetDrawColor(&u8g2, 0);
	u8g2_SetFont(&u8g2, u8g2_font_nokiafc22_tu);
	u8g2_DrawUTF8(&u8g2, 48, 38, "Test");

	scroll_text(&u8g2, 48+soShadow, 45, "This is just a simple test", u8g2_font_5x7_tf); 

	u8g2_SetDrawColor(&u8g2, 1);
}
```

Just before the picture loop, you should copy the original value to the shadow value:

``` C++
volatile int isUpdate = 1;  // init with true so that display is updated during startup

void StartDisp(void *argument)
{
  /* USER CODE BEGIN StartDisp */

  /* Infinite loop */
  for(;;)
  {
          while( isUpdate == 0 )  // do not update the display as long as isUpdate is false
            ;
          /* isUpdate is not 0, so do the update */

          soShadow = so;  // copy the scroll offset

	  u8g2_FirstPage(&u8g2);
	  do {
		  draw_bar();
		  draw_menu();
	  } while (u8g2_NextPage(&u8g2));
          isUpdate = 0; // display update is done, clear the isUpdate flag and wait until it gets set again
  }
  /* USER CODE END StartDisp */
}
```
This approach will keep the scroll offset constant during the display update, avoiding unwanted artefacts on the display.

**Reply by dolence** [reply]
Oli, thank you very much. I had some partial success but I think what I was trying to do isn't exactly what I need. Following the aforementioned example I was able to scroll the string pixel by pixel over the screen. What I really need is to scroll a large string inside a small area of the screen.

I'm trying to draw a menu where the selected option long description is shown beneath the option itself but contained to the area designed to that option. One solution that come to my mind right now is to draw a substring containing part of the full string and update this string by shifting a char to the left as one would do with a character based LCD module. I guess it would work but maybe not as smooth. What you  think?

![image](https://user-images.githubusercontent.com/5223003/224807125-eeecf2c3-21a2-4f26-a2cd-b8c12cd60f74.png)

**Comment by olikraus** [comment, marked as answer]
> to scroll a large string

Long strings are better supported if u8g2 is put into 16 bit mode (see FAQ, https://github.com/olikraus/u8g2/blob/master/doc/faq.txt).

> inside a small area of the screen

Use: https://github.com/olikraus/u8g2/wiki/u8g2reference#setclipwindow

**Reply by dolence** [reply]
Thank you! This is exactly what I needed! 

![image](https://user-images.githubusercontent.com/5223003/225334245-c49a79df-8fca-45e2-a0c7-e5af7fa34df8.png)
